File: plugins/rider/src/main/kotlin/com/shortsleevestudio/gamescript/database/PostgresConnection.kt
package com.shortsleevestudio.gamescript.database

import com.github.jasync.sql.db.Connection
import com.github.jasync.sql.db.QueryResult
import com.github.jasync.sql.db.pool.ConnectionPool
import com.github.jasync.sql.db.postgresql.PostgreSQLConnection
import com.github.jasync.sql.db.postgresql.PostgreSQLConnectionBuilder
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.future.await
import kotlinx.coroutines.withContext
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger

/**
 * PostgreSQL database connection using jasync-sql.
 * Provides async queries, transactions, and LISTEN/NOTIFY support.
 */
class PostgresConnection(
    private val host: String,
    private val port: Int,
    private val database: String,
    private val user: String,
    private val password: String?
) {

    private var pool: ConnectionPool<PostgreSQLConnection>? = null
    private var listenPool: ConnectionPool<PostgreSQLConnection>? = null
    private var listenConnection: PostgreSQLConnection? = null

    private val nextTransactionId = AtomicInteger(1)
    private val activeTransactions = ConcurrentHashMap<String, PostgreSQLConnection>()

    /**
     * Connect to the PostgreSQL database.
     */
    @Suppress("UNCHECKED_CAST")
    suspend fun connect() = withContext(Dispatchers.IO) {
        pool = PostgreSQLConnectionBuilder.createConnectionPool {
            this.host = this@PostgresConnection.host
            this.port = this@PostgresConnection.port
            this.database = this@PostgresConnection.database
            this.username = this@PostgresConnection.user
            this.password = this@PostgresConnection.password
            this.maxActiveConnections = 10
            this.maxIdleTime = 60_000
            this.maxPendingQueries = 100
        } as ConnectionPool<PostgreSQLConnection>

        // Test connection
        pool!!.connect().await()
    }

    /**
     * Close all connections.
     */
    fun close() {
        try {
            // Return listen connection to its pool before disconnecting
            listenConnection?.let { conn ->
                listenPool?.giveBack(conn)
            }
            listenConnection = null
            // Disconnect the listen pool (this properly releases server resources)
            listenPool?.disconnect()
            listenPool = null
            // Disconnect main pool
            pool?.disconnect()
            pool = null
            activeTransactions.clear()
        } catch (_: Exception) {
            // Ignore close errors
        }
    }

    /**
     * Execute a SELECT query.
     * @param connectionId Optional transaction connection ID. If provided, uses that connection.
     */
    suspend fun query(sql: String, params: List<Any?>? = null, connectionId: String? = null): List<Map<String, Any?>> =
        withContext(Dispatchers.IO) {
            val conn = getConnection(connectionId)

            val result = if (params.isNullOrEmpty()) {
                conn.sendQuery(sql).await()
            } else {
                conn.sendPreparedStatement(sql, params).await()
            }

            queryResultToList(result)
        }

    /**
     * Execute an INSERT/UPDATE/DELETE statement.
     * @param connectionId Optional transaction connection ID. If provided, uses that connection.
     */
    suspend fun run(sql: String, params: List<Any?>? = null, returning: Boolean = false, connectionId: String? = null): RunResult =
        withContext(Dispatchers.IO) {
            val conn = getConnection(connectionId)

            val result = if (params.isNullOrEmpty()) {
                conn.sendQuery(sql).await()
            } else {
                conn.sendPreparedStatement(sql, params).await()
            }

            val rows = if (returning) {
                queryResultToList(result)
            } else {
                emptyList()
            }

            val lastId = rows.firstOrNull()?.get("id") as? Long
            RunResult(lastId, result.rowsAffected, rows)
        }

    /**
     * Execute a DDL statement.
     * @param connectionId Optional transaction connection ID. If provided, uses that connection.
     */
    suspend fun exec(sql: String, connectionId: String? = null) = withContext(Dispatchers.IO) {
        val conn = getConnection(connectionId)
        conn.sendQuery(sql).await()
    }

    /**
     * Get a connection - either from a transaction or from the pool.
     * Returns the common Connection interface which both PostgreSQLConnection
     * and ConnectionPool implement, allowing the pool to handle connection
     * borrowing/returning automatically for non-transactional queries.
     */
    private fun getConnection(connectionId: String?): Connection {
        if (connectionId != null) {
            return activeTransactions[connectionId]
                ?: throw IllegalStateException("Transaction $connectionId not found")
        }
        return pool ?: throw IllegalStateException("Not connected")
    }

    /**
     * Begin a new transaction.
     * Gets a dedicated connection from the pool for this transaction.
     */
    suspend fun beginTransaction(): String = withContext(Dispatchers.IO) {
        val connectionPool = pool ?: throw IllegalStateException("Not connected")
        val id = "tx_${nextTransactionId.getAndIncrement()}"

        // Get a dedicated connection from the pool for this transaction
        val dedicatedConn = connectionPool.take().await()
        dedicatedConn.sendQuery("BEGIN").await()
        activeTransactions[id] = dedicatedConn
        id
    }

    /**
     * Commit a transaction.
     * Returns the dedicated connection back to the pool.
     */
    suspend fun commitTransaction(connectionId: String) = withContext(Dispatchers.IO) {
        val conn = activeTransactions.remove(connectionId)
            ?: throw IllegalStateException("Transaction $connectionId not found")

        try {
            conn.sendQuery("COMMIT").await()
        } finally {
            // Return connection to pool
            pool?.giveBack(conn)
        }
    }

    /**
     * Rollback a transaction.
     * Returns the dedicated connection back to the pool.
     */
    suspend fun rollbackTransaction(connectionId: String) = withContext(Dispatchers.IO) {
        val conn = activeTransactions.remove(connectionId)
            ?: throw IllegalStateException("Transaction $connectionId not found")

        try {
            conn.sendQuery("ROLLBACK").await()
        } finally {
            // Return connection to pool
            pool?.giveBack(conn)
        }
    }

    /**
     * Start listening for notifications on a channel.
     * Uses a dedicated connection for LISTEN/NOTIFY.
     */
    @Suppress("UNCHECKED_CAST")
    suspend fun startListening(channel: String, onNotification: (String) -> Unit) =
        withContext(Dispatchers.IO) {
            // Create dedicated pool for listening (must store pool reference for proper cleanup)
            listenPool = PostgreSQLConnectionBuilder.createConnectionPool {
                this.host = this@PostgresConnection.host
                this.port = this@PostgresConnection.port
                this.database = this@PostgresConnection.database
                this.username = this@PostgresConnection.user
                this.password = this@PostgresConnection.password
                this.maxActiveConnections = 1
            } as ConnectionPool<PostgreSQLConnection>

            // Get a dedicated connection from the pool
            listenConnection = listenPool!!.take().await()

            // Register notification listener
            listenConnection!!.registerNotifyListener { notification ->
                if (notification.channel == channel) {
                    onNotification(notification.payload)
                }
            }

            // Start listening
            listenConnection!!.sendQuery("LISTEN $channel").await()
        }

    /**
     * Stop listening for notifications.
     */
    suspend fun stopListening(channel: String) = withContext(Dispatchers.IO) {
        listenConnection?.sendQuery("UNLISTEN $channel")?.await()
        // Return connection to pool before disconnecting the pool
        listenConnection?.let { conn ->
            listenPool?.giveBack(conn)
        }
        listenConnection = null
        // Disconnect the pool to release server resources
        listenPool?.disconnect()
        listenPool = null
    }

    private fun queryResultToList(result: QueryResult): List<Map<String, Any?>> {
        val rows = result.rows
        if (rows.isEmpty()) return emptyList()

        val columns = rows.columnNames()
        return rows.map { row ->
            columns.mapIndexed { index, column ->
                column to row[index]
            }.toMap()
        }
    }
}

File: plugins/dotnet/visual-studio/GameScript.VisualStudio/Database/PostgresConnection.cs
using Npgsql;
using System.Collections.Concurrent;

namespace GameScript.VisualStudio;

/// <summary>
/// PostgreSQL database connection using Npgsql.
/// Provides async queries, transactions, and LISTEN/NOTIFY support.
/// </summary>
public sealed class PostgresConnection : IDisposable
{
    private readonly string _host;
    private readonly int _port;
    private readonly string _database;
    private readonly string _user;
    private readonly string? _password;

    private NpgsqlDataSource? _dataSource;
    private NpgsqlConnection? _listenConnection;
    private CancellationTokenSource? _listenCancellation;
    private int _nextTransactionId = 1;
    private readonly ConcurrentDictionary<string, NpgsqlConnection> _activeTransactions = new();
    private Action<string>? _notificationCallback;

    public PostgresConnection(
        string host,
        int port,
        string database,
        string user,
        string? password)
    {
        _host = host;
        _port = port;
        _database = database;
        _user = user;
        _password = password;
    }

    /// <summary>
    /// Connect to the PostgreSQL database.
    /// </summary>
    public async Task ConnectAsync(CancellationToken cancellationToken = default)
    {
        var connectionString = new NpgsqlConnectionStringBuilder
        {
            Host = _host,
            Port = _port,
            Database = _database,
            Username = _user,
            Password = _password,
            MaxPoolSize = 10,
            MinPoolSize = 1,
            ConnectionIdleLifetime = 60
        }.ToString();

        _dataSource = NpgsqlDataSource.Create(connectionString);

        // Test connection
        await using var conn = await _dataSource.OpenConnectionAsync(cancellationToken);
    }

    /// <summary>
    /// Close all connections.
    /// </summary>
    public async Task CloseAsync()
    {
        try
        {
            if (_listenConnection != null)
            {
                await _listenConnection.CloseAsync();
                await _listenConnection.DisposeAsync();
                _listenConnection = null;
            }

            // Close transaction connections
            foreach (var (_, conn) in _activeTransactions)
            {
                await conn.CloseAsync();
                await conn.DisposeAsync();
            }
            _activeTransactions.Clear();

            if (_dataSource != null)
            {
                await _dataSource.DisposeAsync();
                _dataSource = null;
            }
        }
        catch
        {
            // Ignore close errors
        }
    }

    /// <summary>
    /// Execute a SELECT query.
    /// </summary>
    public async Task<List<Dictionary<string, object?>>> QueryAsync(
        string sql,
        List<object?>? parameters = null,
        string? connectionId = null,
        CancellationToken cancellationToken = default)
    {
        var conn = await GetConnectionAsync(connectionId, cancellationToken);

        try
        {
            await using var cmd = new NpgsqlCommand(sql, conn);
            BindParameters(cmd, parameters);

            await using var reader = await cmd.ExecuteReaderAsync(cancellationToken);
            return await ReaderToListAsync(reader, cancellationToken);
        }
        finally
        {
            // Only close if we opened it (not a transaction connection)
            if (connectionId == null)
            {
                await conn.CloseAsync();
            }
        }
    }

    /// <summary>
    /// Execute an INSERT/UPDATE/DELETE statement.
    /// </summary>
    public async Task<RunResult> RunAsync(
        string sql,
        List<object?>? parameters = null,
        bool returning = false,
        string? connectionId = null,
        CancellationToken cancellationToken = default)
    {
        var conn = await GetConnectionAsync(connectionId, cancellationToken);

        try
        {
            await using var cmd = new NpgsqlCommand(sql, conn);
            BindParameters(cmd, parameters);

            if (returning)
            {
                await using var reader = await cmd.ExecuteReaderAsync(cancellationToken);
                var rows = await ReaderToListAsync(reader, cancellationToken);
                var lastId = rows.FirstOrDefault()?.TryGetValue("id", out var id) == true ? Convert.ToInt64(id) : (long?)null;
                return new RunResult
                {
                    LastId = lastId,
                    Changes = rows.Count,
                    Rows = rows
                };
            }
            else
            {
                var changes = await cmd.ExecuteNonQueryAsync(cancellationToken);
                return new RunResult
                {
                    LastId = null,
                    Changes = changes,
                    Rows = new List<Dictionary<string, object?>>()
                };
            }
        }
        finally
        {
            if (connectionId == null)
            {
                await conn.CloseAsync();
            }
        }
    }

    /// <summary>
    /// Execute a DDL statement.
    /// </summary>
    public async Task ExecAsync(
        string sql,
        string? connectionId = null,
        CancellationToken cancellationToken = default)
    {
        var conn = await GetConnectionAsync(connectionId, cancellationToken);

        try
        {
            await using var cmd = new NpgsqlCommand(sql, conn);
            await cmd.ExecuteNonQueryAsync(cancellationToken);
        }
        finally
        {
            if (connectionId == null)
            {
                await conn.CloseAsync();
            }
        }
    }

    /// <summary>
    /// Get a connection - either from a transaction or from the pool.
    /// </summary>
    private async Task<NpgsqlConnection> GetConnectionAsync(
        string? connectionId,
        CancellationToken cancellationToken)
    {
        if (connectionId != null)
        {
            if (_activeTransactions.TryGetValue(connectionId, out var txConn))
            {
                return txConn;
            }
            throw new InvalidOperationException($"Transaction {connectionId} not found");
        }

        if (_dataSource == null)
        {
            throw new InvalidOperationException("Not connected");
        }

        return await _dataSource.OpenConnectionAsync(cancellationToken);
    }

    /// <summary>
    /// Begin a new transaction.
    /// Gets a dedicated connection from the pool for this transaction.
    /// </summary>
    public async Task<string> BeginTransactionAsync(CancellationToken cancellationToken = default)
    {
        if (_dataSource == null)
        {
            throw new InvalidOperationException("Not connected");
        }

        var id = $"tx_{Interlocked.Increment(ref _nextTransactionId)}";
        var conn = await _dataSource.OpenConnectionAsync(cancellationToken);

        try
        {
            await using var cmd = new NpgsqlCommand("BEGIN", conn);
            await cmd.ExecuteNonQueryAsync(cancellationToken);

            _activeTransactions[id] = conn;
            return id;
        }
        catch
        {
            await conn.CloseAsync();
            throw;
        }
    }

    /// <summary>
    /// Commit a transaction.
    /// Returns the dedicated connection back to the pool.
    /// </summary>
    public async Task CommitTransactionAsync(string connectionId, CancellationToken cancellationToken = default)
    {
        if (!_activeTransactions.TryRemove(connectionId, out var conn))
        {
            throw new InvalidOperationException($"Transaction {connectionId} not found");
        }

        try
        {
            await using var cmd = new NpgsqlCommand("COMMIT", conn);
            await cmd.ExecuteNonQueryAsync(cancellationToken);
        }
        finally
        {
            await conn.CloseAsync();
        }
    }

    /// <summary>
    /// Rollback a transaction.
    /// Returns the dedicated connection back to the pool.
    /// </summary>
    public async Task RollbackTransactionAsync(string connectionId, CancellationToken cancellationToken = default)
    {
        if (!_activeTransactions.TryRemove(connectionId, out var conn))
        {
            throw new InvalidOperationException($"Transaction {connectionId} not found");
        }

        try
        {
            await using var cmd = new NpgsqlCommand("ROLLBACK", conn);
            await cmd.ExecuteNonQueryAsync(cancellationToken);
        }
        finally
        {
            await conn.CloseAsync();
        }
    }

    /// <summary>
    /// Start listening for notifications on a channel.
    /// Uses a dedicated connection for LISTEN/NOTIFY.
    /// </summary>
    public async Task StartListeningAsync(
        string channel,
        Action<string> onNotification,
        CancellationToken cancellationToken = default)
    {
        if (_dataSource == null)
        {
            throw new InvalidOperationException("Not connected");
        }

        // Cancel any existing listener
        _listenCancellation?.Cancel();
        _listenCancellation?.Dispose();
        _listenCancellation = new CancellationTokenSource();

        _notificationCallback = onNotification;
        _listenConnection = await _dataSource.OpenConnectionAsync(cancellationToken);

        // Register notification handler
        _listenConnection.Notification += (_, e) =>
        {
            if (e.Channel == channel)
            {
                _notificationCallback?.Invoke(e.Payload);
            }
        };

        // Start listening
        await using var cmd = new NpgsqlCommand($"LISTEN {channel}", _listenConnection);
        await cmd.ExecuteNonQueryAsync(cancellationToken);

        // Capture the cancellation token for the background task
        CancellationToken listenToken = _listenCancellation.Token;

        // Start background task to wait for notifications
        _ = Task.Run(async () =>
        {
            try
            {
                while (!listenToken.IsCancellationRequested &&
                       _listenConnection != null &&
                       _listenConnection.State == System.Data.ConnectionState.Open)
                {
                    await _listenConnection.WaitAsync(listenToken);
                }
            }
            catch (OperationCanceledException)
            {
                // Expected when stopping
            }
            catch
            {
                // Connection closed
            }
        }, listenToken);
    }

    /// <summary>
    /// Stop listening for notifications.
    /// </summary>
    public async Task StopListeningAsync(string channel, CancellationToken cancellationToken = default)
    {
        // Cancel the background listener task
        _listenCancellation?.Cancel();
        _listenCancellation?.Dispose();
        _listenCancellation = null;

        if (_listenConnection != null)
        {
            try
            {
                await using var cmd = new NpgsqlCommand($"UNLISTEN {channel}", _listenConnection);
                await cmd.ExecuteNonQueryAsync(cancellationToken);
            }
            catch
            {
                // Ignore errors during cleanup
            }

            await _listenConnection.CloseAsync();
            await _listenConnection.DisposeAsync();
            _listenConnection = null;
        }

        _notificationCallback = null;
    }

    private static void BindParameters(NpgsqlCommand cmd, List<object?>? parameters)
    {
        if (parameters == null) return;

        for (int i = 0; i < parameters.Count; i++)
        {
            var value = parameters[i];
            cmd.Parameters.AddWithValue($"${i + 1}", value ?? DBNull.Value);
        }
    }

    private static async Task<List<Dictionary<string, object?>>> ReaderToListAsync(
        NpgsqlDataReader reader,
        CancellationToken cancellationToken)
    {
        var results = new List<Dictionary<string, object?>>();

        while (await reader.ReadAsync(cancellationToken))
        {
            var row = new Dictionary<string, object?>();
            for (int i = 0; i < reader.FieldCount; i++)
            {
                var columnName = reader.GetName(i);
                var value = reader.IsDBNull(i) ? null : reader.GetValue(i);
                row[columnName] = value;
            }
            results.Add(row);
        }

        return results;
    }

    public void Dispose()
    {
        _listenCancellation?.Cancel();
        _listenCancellation?.Dispose();
        _listenConnection?.Dispose();
        foreach (var (_, conn) in _activeTransactions)
        {
            conn.Dispose();
        }
        _activeTransactions.Clear();
        _dataSource?.Dispose();
    }
}
